File: test_mcp_client.py
#!/usr/bin/env python3
"""
Simple MCP client to test fern-mycelium MCP server functionality.
This script demonstrates how an AI agent would query the test context.
"""

import json
import requests
import sys

def test_graphql_query(query, variables=None):
    """Test a GraphQL query against the fern-mycelium server."""
    url = "http://localhost:8081/query"
    payload = {
        "query": query
    }
    if variables:
        payload["variables"] = variables
    
    try:
        response = requests.post(url, json=payload, headers={'Content-Type': 'application/json'})
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"❌ Error querying GraphQL: {e}")
        return None

def main():
    print("🧪 Testing fern-mycelium MCP Server")
    print("=" * 50)
    
    # Test 1: Health check
    print("\n1. Testing Health Check...")
    try:
        response = requests.get("http://localhost:8081/healthz")
        if response.status_code == 200:
            health_data = response.json()
            print(f"✅ Health: {health_data['status']} - {health_data['message']}")
        else:
            print(f"❌ Health check failed: {response.status_code}")
            return
    except Exception as e:
        print(f"❌ Health check error: {e}")
        return
    
    # Test 2: Query flaky tests
    print("\n2. Testing Flaky Test Detection...")
    flaky_query = """
    {
        flakyTests(limit: 10, projectID: "MCP Server Tests") {
            testID
            testName
            passRate
            failureRate
            runCount
            lastFailure
        }
    }
    """
    
    result = test_graphql_query(flaky_query)
    if result and 'data' in result and 'flakyTests' in result['data']:
        flaky_tests = result['data']['flakyTests']
        print(f"✅ Found {len(flaky_tests)} tests with flaky behavior:")
        
        for test in flaky_tests:
            status = "🔴 FLAKY" if test['failureRate'] > 0 else "🟢 STABLE"
            print(f"   {status} {test['testName']}")
            print(f"      - Pass Rate: {test['passRate']:.1%}")
            print(f"      - Failure Rate: {test['failureRate']:.1%}")
            print(f"      - Total Runs: {test['runCount']}")
            if test['lastFailure']:
                print(f"      - Last Failure: {test['lastFailure']}")
            print()
    else:
        print(f"❌ Failed to query flaky tests: {result}")
        return
    
    # Test 3: Demonstrate AI Agent Use Case
    print("3. AI Agent Analysis Simulation...")
    print("🤖 Agent: Analyzing test stability patterns...")
    
    high_risk_tests = [test for test in flaky_tests if test['failureRate'] > 0.3]
    moderate_risk_tests = [test for test in flaky_tests if 0 < test['failureRate'] <= 0.3]
    stable_tests = [test for test in flaky_tests if test['failureRate'] == 0]
    
    print(f"\n📊 Test Stability Analysis:")
    print(f"   🔴 High Risk (>30% failure): {len(high_risk_tests)} tests")
    print(f"   🟡 Moderate Risk (1-30% failure): {len(moderate_risk_tests)} tests")
    print(f"   🟢 Stable (0% failure): {len(stable_tests)} tests")
    
    if high_risk_tests:
        print(f"\n🚨 Agent Recommendation: Immediate attention needed for:")
        for test in high_risk_tests:
            print(f"   - {test['testName']} ({test['failureRate']:.1%} failure rate)")
    
    if moderate_risk_tests:
        print(f"\n⚠️  Agent Recommendation: Monitor closely:")
        for test in moderate_risk_tests:
            print(f"   - {test['testName']} ({test['failureRate']:.1%} failure rate)")
    
    print("\n✅ MCP Server Test Complete!")
    print("🎯 The fern-mycelium MCP server successfully provides test intelligence context")
    print("   that AI agents can use for automated test analysis and recommendations!")

if __name__ == "__main__":
    main()

File: tools/test_mcp_client.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// GraphQL request and response structures
type GraphQLRequest struct {
	Query     string                 `json:"query"`
	Variables map[string]interface{} `json:"variables,omitempty"`
}

type GraphQLResponse struct {
	Data   map[string]interface{} `json:"data,omitempty"`
	Errors []GraphQLError         `json:"errors,omitempty"`
}

type GraphQLError struct {
	Message string        `json:"message"`
	Path    []interface{} `json:"path,omitempty"`
}

type HealthResponse struct {
	Status  string `json:"status"`
	Message string `json:"message"`
}

type FlakyTest struct {
	TestID      string  `json:"testID"`
	TestName    string  `json:"testName"`
	PassRate    float64 `json:"passRate"`
	FailureRate float64 `json:"failureRate"`
	RunCount    int     `json:"runCount"`
	LastFailure *string `json:"lastFailure"`
}

const (
	baseURL = "http://localhost:8081"
)

func main() {
	fmt.Println("🧪 Testing fern-mycelium MCP Server")
	fmt.Println("==================================================")

	// Test 1: Health check
	fmt.Println("\n1. Testing Health Check...")
	if !testHealthCheck() {
		os.Exit(1)
	}

	// Test 2: Query flaky tests
	fmt.Println("\n2. Testing Flaky Test Detection...")
	flakyTests, ok := testFlakyTestQuery()
	if !ok {
		os.Exit(1)
	}

	// Test 3: Demonstrate AI Agent Use Case
	fmt.Println("\n3. AI Agent Analysis Simulation...")
	simulateAgentAnalysis(flakyTests)

	fmt.Println("\n✅ MCP Server Test Complete!")
	fmt.Println("🎯 The fern-mycelium MCP server successfully provides test intelligence context")
	fmt.Println("   that AI agents can use for automated test analysis and recommendations!")
}

func testHealthCheck() bool {
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(baseURL + "/healthz")
	if err != nil {
		fmt.Printf("❌ Health check error: %v\n", err)
		return false
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		fmt.Printf("❌ Health check failed: %d\n", resp.StatusCode)
		return false
	}

	var health HealthResponse
	if err := json.NewDecoder(resp.Body).Decode(&health); err != nil {
		fmt.Printf("❌ Failed to decode health response: %v\n", err)
		return false
	}

	fmt.Printf("✅ Health: %s - %s\n", health.Status, health.Message)
	return true
}

func testFlakyTestQuery() ([]FlakyTest, bool) {
	query := `{
		flakyTests(limit: 10, projectID: "MCP Server Tests") {
			testID
			testName
			passRate
			failureRate
			runCount
			lastFailure
		}
	}`

	result, err := executeGraphQLQuery(query, nil)
	if err != nil {
		fmt.Printf("❌ Failed to query flaky tests: %v\n", err)
		return nil, false
	}

	if len(result.Errors) > 0 {
		fmt.Printf("❌ GraphQL errors: %v\n", result.Errors)
		return nil, false
	}

	// Parse flaky tests from response
	flakyTestsData, ok := result.Data["flakyTests"].([]interface{})
	if !ok {
		fmt.Printf("❌ Invalid flaky tests data format\n")
		return nil, false
	}

	var flakyTests []FlakyTest
	for _, item := range flakyTestsData {
		testMap, ok := item.(map[string]interface{})
		if !ok {
			continue
		}

		test := FlakyTest{
			TestID:      getString(testMap, "testID"),
			TestName:    getString(testMap, "testName"),
			PassRate:    getFloat64(testMap, "passRate"),
			FailureRate: getFloat64(testMap, "failureRate"),
			RunCount:    int(getFloat64(testMap, "runCount")),
		}

		if lastFailure, exists := testMap["lastFailure"]; exists && lastFailure != nil {
			if failureStr, ok := lastFailure.(string); ok {
				test.LastFailure = &failureStr
			}
		}

		flakyTests = append(flakyTests, test)
	}

	fmt.Printf("✅ Found %d tests with flaky behavior:\n", len(flakyTests))
	for _, test := range flakyTests {
		status := "🔴 FLAKY"
		if test.FailureRate == 0 {
			status = "🟢 STABLE"
		}

		fmt.Printf("   %s %s\n", status, test.TestName)
		fmt.Printf("      - Pass Rate: %.1f%%\n", test.PassRate*100)
		fmt.Printf("      - Failure Rate: %.1f%%\n", test.FailureRate*100)
		fmt.Printf("      - Total Runs: %d\n", test.RunCount)
		if test.LastFailure != nil {
			fmt.Printf("      - Last Failure: %s\n", *test.LastFailure)
		}
		fmt.Println()
	}

	return flakyTests, true
}

func simulateAgentAnalysis(flakyTests []FlakyTest) {
	fmt.Println("🤖 Agent: Analyzing test stability patterns...")

	var highRisk, moderateRisk, stable []FlakyTest

	for _, test := range flakyTests {
		if test.FailureRate > 0.3 {
			highRisk = append(highRisk, test)
		} else if test.FailureRate > 0 {
			moderateRisk = append(moderateRisk, test)
		} else {
			stable = append(stable, test)
		}
	}

	fmt.Printf("\n📊 Test Stability Analysis:\n")
	fmt.Printf("   🔴 High Risk (>30%% failure): %d tests\n", len(highRisk))
	fmt.Printf("   🟡 Moderate Risk (1-30%% failure): %d tests\n", len(moderateRisk))
	fmt.Printf("   🟢 Stable (0%% failure): %d tests\n", len(stable))

	if len(highRisk) > 0 {
		fmt.Printf("\n🚨 Agent Recommendation: Immediate attention needed for:\n")
		for _, test := range highRisk {
			fmt.Printf("   - %s (%.1f%% failure rate)\n", test.TestName, test.FailureRate*100)
		}
	}

	if len(moderateRisk) > 0 {
		fmt.Printf("\n⚠️  Agent Recommendation: Monitor closely:\n")
		for _, test := range moderateRisk {
			fmt.Printf("   - %s (%.1f%% failure rate)\n", test.TestName, test.FailureRate*100)
		}
	}
}

func executeGraphQLQuery(query string, variables map[string]interface{}) (*GraphQLResponse, error) {
	reqBody := GraphQLRequest{
		Query:     query,
		Variables: variables,
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Post(baseURL+"/query", "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP error %d: %s", resp.StatusCode, string(body))
	}

	var result GraphQLResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &result, nil
}

// Helper functions for type conversion
func getString(m map[string]interface{}, key string) string {
	if val, ok := m[key].(string); ok {
		return val
	}
	return ""
}

func getFloat64(m map[string]interface{}, key string) float64 {
	if val, ok := m[key].(float64); ok {
		return val
	}
	return 0
}